lib/swaps/Swaps.ts

```diff
@@ -961,42 +1013,60 @@ class Swaps extends EventEmitter {
         deal.rPreimage = await swapClient.sendPayment(deal);
-        return deal.rPreimage;
       } catch (err) {
-        // the payment failed but we are unsure of its final status, so we fail
-        // the deal and assign the payment to be checked in swap recovery.
-        // we don't remove our incoming invoice because we are not yet certain
-        // whether our outgoing payment can be claimed by the taker or not
-        switch (err.code) {
-          case errorCodes.FINAL_PAYMENT_ERROR:
-            await this.failDeal({
-              deal,
-              peer,
-              failedCurrency: deal.takerCurrency,
-              failureReason: SwapFailureReason.SendPaymentFailure,
-              errorMessage: err.message,
-            });
-            break;
-          case errorCodes.PAYMENT_REJECTED:
-            await this.failDeal({
-              deal,
-              failureReason: SwapFailureReason.PaymentRejected,
-              errorMessage: err.message,
-            });
-            break;
-          default:
-            await this.failDeal({
-              deal,
-              peer,
-              failedCurrency: deal.takerCurrency,
-              failureReason: SwapFailureReason.UnknownError,
-              errorMessage: err.message,
-            });
-            break;
-        }
+        this.logger.debug(`sendPayment in resolveHash failed due to ${err.message}`);
+
+        const failDeal = async (paymentState: PaymentState) => {
+          switch (err.code) {
+            case errorCodes.FINAL_PAYMENT_ERROR:
+              await this.failDeal({
+                deal,
+                peer,
+                paymentState,
+                failedCurrency: deal.takerCurrency,
+                failureReason: SwapFailureReason.SendPaymentFailure,
+                errorMessage: err.message,
+              });
+              break;
+            case errorCodes.PAYMENT_REJECTED:
+              await this.failDeal({
+                deal,
+                paymentState,
+                failureReason: SwapFailureReason.PaymentRejected,
+                errorMessage: err.message,
+              });
+              break;
+            default:
+              await this.failDeal({
+                deal,
+                peer,
+                paymentState,
+                failedCurrency: deal.takerCurrency,
+                failureReason: SwapFailureReason.UnknownError,
+                errorMessage: err.message,
+              });
+              break;
+          }
+        };
 
-        throw err;
+        // the payment failed but we first double check its final status, so we
+        // only fail the deal when we know our payment won't go through. otherwise
+        // we extract the preimage if the payment went through in spite of the error
+        // or we fail the deal and go to SwapRecovery if it's still pending
+        const paymentStatus = await swapClient.lookupPayment(rHash, deal.makerCurrency, deal.destination);
```

Comment:
Think this should be `deal.takerCurrency` as maker's payment is in taker's currency.

Reply:
Good catch, you are right. It's not a parameter that's used for LND but it wouldn't work for connext.